### ISSUE TITLE
meson build -Db_coverage=true fails on harfbuzz when meson is upgraded to 0.55

### ISSUE THREAD

**Issue by ebraminio** [mesonbuild/meson#7437]
`meson build -Db_coverage=true` fails on harfbuzz when meson is upgraded to 0.55 and gives,  https://github.com/harfbuzz/harfbuzz/runs/863065338

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/mesonbuild/mesonmain.py", line 131, in run
    return options.run_func(options)
  File "/usr/local/lib/python3.8/dist-packages/mesonbuild/msetup.py", line 245, in run
    app.generate()
  File "/usr/local/lib/python3.8/dist-packages/mesonbuild/msetup.py", line 159, in generate
    self._generate(env)
  File "/usr/local/lib/python3.8/dist-packages/mesonbuild/msetup.py", line 215, in _generate
    intr.backend.generate()
  File "/usr/local/lib/python3.8/dist-packages/mesonbuild/backend/ninjabackend.py", line 518, in generate
    self.generate_coverage_rules()
  File "/usr/local/lib/python3.8/dist-packages/mesonbuild/backend/ninjabackend.py", line 991, in generate_coverage_rules
    self.generate_coverage_command(e, [])
  File "/usr/local/lib/python3.8/dist-packages/mesonbuild/backend/ninjabackend.py", line 975, in generate_coverage_command
    for compiler in target.compilers.values():
AttributeError: 'CustomTarget' object has no attribute 'compilers'
```

I rechecked manually that wasn't the case with the prior version as this happened with the meson upgrade on pip.

**To Reproduce**
`git clone https://github.com/harfbuzz/harfbuzz && cd harfbuzz && meson build -Db_coverage=true`, doesn't happen before 0.55 but happens after it

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? No
* what operating system, Ubuntu latest https://github.com/harfbuzz/harfbuzz/runs/863065338 which runs https://github.com/harfbuzz/harfbuzz/blob/master/.github/workflows/linux-ci.yml#L11 but happened on another bot with 20.04 also and locally on Manjaro/ArchLinux.
* what Python version are you using, 3.8.2
* what `meson --version`, 0.55.0
* what `ninja --version` if it's a Ninja build, 1.10.0

**Comment by ebraminio** [comment]
(Just to not forget, likewise we have an gobject-instrospection build issue lhttps://ci.appveyor.com/project/harfbuzz/harfbuzz/builds/34050794/job/gpj95udn3owvw817#L177 which will file in separately after more investigation)

**Comment by lazka** [comment]
The problem is this line `g_ir_scanner=${bindir}/g-ir-scanner` in `gobject-introspection-1.0.pc`. Meson expands it and then throws it out because it's missing a file extension (in _search_windows_special_cases()), and then never checks if it succeeded.

So, the .pc file in MSYS2 needs to be fixed and append '.exe' there.

**Comment by georgev93** [comment]
PR #7411 addresses this best

**Comment by alex-tee** [comment]
I have the same issue: https://travis-ci.org/github/zrythm/zrythm/jobs/707588823

**Comment by phillipjohnston** [comment]
Also seeing this on my projects.

**Comment by jpakkane** [comment]
Can you test whether #7458 or #7411 fix it for you?

**Comment by ebraminio** [comment]
Sure,
```
$ git clone https://github.com/harfbuzz/harfbuzz && cd harfbuzz
$ pip install git+https://github.com/mesonbuild/meson
$ meson build -Db_coverage=true 
(fails)
$ pip install git+https://github.com/marcelhollerbach/meson@bugfix
$ meson build -Db_coverage=true 
(fails)
$ pip install git+https://github.com/phongt/meson@fix_coverage_run_target_problem
$ meson build -Db_coverage=true
(fails)
```

Some maybe something is wrong with my testing?

I should've put a bot running master branch of meson in harfbuzz CI to catch this, and in fact I've done that just not on the specific bot that had -Db_coverage :/ If you got sometime feel free to have a look at the meson port raise issues you see in some issue before the release as we are going to make meson our main build on next release https://github.com/harfbuzz/harfbuzz/pull/2556/files . Thanks!

**Comment by jpakkane** [comment]
I tried the fix in #7411 by hand in my main checkout by hand it it does fix the issue. So something is wonky with your setup, maybe it does not switch to the named branch before running the test?

> have a look at our meson port and raise issues you see in some file before the release 

These struck me immediately on a build:

```
cc1: warning: command line option ‘-fno-rtti’ is valid for C++/D/ObjC++ but not for C
cc1: warning: command line option ‘-fno-threadsafe-statics’ is valid for C++/ObjC++ but not for C
cc1: warning: command line option ‘-fvisibility-inlines-hidden’ is valid for C++/ObjC++ but not for C
cc1: warning: command line option ‘-fno-rtti’ is valid for C++/D/ObjC++ but not for C
cc1: warning: command line option ‘-fno-threadsafe-statics’ is valid for C++/ObjC++ but not for C
cc1: warning: command line option ‘-fvisibility-inlines-hidden’ is valid for C++/ObjC++ but not for C
```

There are builtin options for most of these. You should use those instead. For example using `gnu_symbol_visibility` instead of adding `-fvisibility-xxx` command line arguments by hand.

**Comment by alex-tee** [comment]
I just tested both #7458 and #7411 , they seem to fix the issue.
I tried 0.55.0 again just to re-check and that fails.

**Comment by eli-schwartz** [comment]
Fixed in master by #7458

**Comment by ebraminio** [comment]
To test the fix and early catch regression we are now testing meson's master in one of our bots https://github.com/harfbuzz/harfbuzz/commit/7bab6087bf1b7809ce05a9fb07a40d5708e4ff5d so as the change looks fine am closing this. Thanks!
